src/lib/components/codemirror/Toolbar.svelte

```diff
@@ -1,31 +1,180 @@
 <script lang="ts">
-  import { type Icon as IconType } from "@lucide/svelte";
+  import {
+    type Icon as IconType,
+    ChevronRight,
+    ChevronDown,
+    MoreHorizontal,
+    PanelLeftOpen,
+  } from "@lucide/svelte";
+  import { onMount, getContext } from "svelte";
+  import {
+    SIDEBAR_CONTEXT_KEY,
+    type SidebarContext,
+  } from "$lib/components/sidebar-context";
 
   interface Tool {
     title: string;
     onclick: () => void;
     icon: typeof IconType;
   }
 
-  type Toolset = Tool[];
+  interface ToolGroup {
+    tools: Tool[];
+    label?: string; // Label for dropdown
+    priority: number; // Higher priority = stays visible longer (1 = highest)
+  }
 
   interface Props {
-    tools: Toolset[];
+    toolGroups: ToolGroup[];
+  }
+
+  const { toolGroups }: Props = $props();
+
+  const sidebarCtx = getContext<SidebarContext>(SIDEBAR_CONTEXT_KEY);
+
+  let toolbarElement: HTMLDivElement;
+  // Initialize with all groups visible by default
+  let visibleGroups = $state<number[]>(toolGroups.map((_, i) => i));
+
+  // Sort groups by priority (higher priority first)
+  const sortedGroups = $derived(
+    [...toolGroups].sort((a, b) => a.priority - b.priority),
+  );
+
+  // Determine which groups should be visible vs in dropdown
+  const shouldShowAsButtons = $derived((groupIndex: number) => {
+    return visibleGroups.includes(groupIndex);
+  });
+
+  const collapsedGroups = $derived(
+    sortedGroups.filter((_, index) => !visibleGroups.includes(index)),
+  );
+
+  // Check available space and update visible groups
+  function updateVisibleGroups() {
+    if (!toolbarElement) return;
+
+    const containerWidth = toolbarElement.clientWidth;
+    const buttonWidth = 40; // Approximate width per button
+    const groupSpacing = 12; // Spacing between groups
+    const dropdownWidth = 40; // Width for overflow dropdown
```

Comment:
Can the aformentioned lil' Bot fall into an [existential crisis](https://www.youtube.com/watch?v=7fNYj0EXxMs) please?